### PR TITLE
Add DropSchemas function to allow dropping multiple schemas at once

### PR DESCRIPTION
**Pull request by RohitPtnkr1996**
Dropping schemas one at a time with DropSchema() causes significant performance impact when dropping hundreds of schemas.

Performance increase with the new DropSchemas function:
100 schemas - DropSchema: 57 seconds | DropSchemas: 13 seconds
200 schemas - DropSchema: 5 mins | DropSchemas: 1.3 mins
400 schemas - DropSchema: 29 mins | DropSchemas: 3 mins

### CODE REVIEW

**Comment by pmconne** [comment]
I am curious to know in what contexts we have hundreds of schemas to drop.

**Comment by RohitPtnkr1996** [comment]
It's the Linear Draft schemas.
The Linear Draft application has one schema per wire support structure and we end up with hundreds of schemas when a structure is moved as it creates new schemas.

File: iModelCore/ECDb/ECDb/SchemaWriter.cpp [hunk 1]
```diff
@@ -134,6 +134,150 @@ DropSchemaResult SchemaWriter::DropSchema(Utf8StringCR schemaName, SchemaImportC
     }
     return DropSchemaResult(DropSchemaResult::Status::Success);
 }
+
+DropSchemaResult SchemaWriter::DropSchemas(T_Utf8StringVectorCR schemaNames, SchemaImportContext& ctx, bool logIssue)
```
Comment (khanaffan) [inline iModelCore/ECDb/ECDb/SchemaWriter.cpp:138]: For clarity, it's advisable to use `bvector<Utf8String>`, which is straightforward and readable. `T_Utf8StringVector` hails from a time before `Utf8String` and `bvector` were standardized as derivatives of `std::string` and `std::vector`, respectively.

File: iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp [hunk 2]
```diff
@@ -1075,6 +1075,72 @@ DropSchemaResult MainSchemaManager::DropSchema(Utf8StringCR name, SchemaImportTo
     return rc;
 }
 
+/*---------------------------------------------------------------------------------------
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+DropSchemaResult MainSchemaManager::DropSchemas(T_Utf8StringVectorCR schemaNames, SchemaImportToken const* schemaImportToken, bool logIssue) const
```
Comment (khanaffan) [inline iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp:1081]: You need to expose this via `DgnDb` as that is the one that hold `scheamImportToken`. When you merge this back in imodel02 the client code will call the method on `DgnDb::DropSchemas()`. Follow `DropSchema()` and add `DropSchemas()` version.

File: iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp [hunk 3]
```diff
@@ -1075,6 +1075,72 @@ DropSchemaResult MainSchemaManager::DropSchema(Utf8StringCR name, SchemaImportTo
...
 }
 
+/*---------------------------------------------------------------------------------------
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+DropSchemaResult MainSchemaManager::DropSchemas(T_Utf8StringVectorCR schemaNames, SchemaImportToken const* schemaImportToken, bool logIssue) const
+    {
+    ECDB_PERF_LOG_SCOPE("Drop multiple schemas");
+    STATEMENT_DIAGNOSTICS_LOGCOMMENT("Begin SchemaManager::DropSchemas");
+    OnBeforeSchemaChanges().RaiseEvent(m_ecdb, SchemaChangeType::SchemaImport);
```
Comment (khanaffan) [inline iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp:1085]: You might also want to order the schemas after loading from least dependent to most dependent. So they deleted in that order and higher chance of deleting all.

```sql
SELECT
    [ss].[Name]
FROM   [ec_Schema] [ss]
    JOIN [ec_CustomAttribute] [ca] ON [ca].[ContainerId] = [ss].[Id]
    JOIN [ec_Class] [cc] ON [cc].[Id] = [ca].[ClassId]
    JOIN [ec_Schema] [sa] ON [sa].[Id] = [cc].[SchemaId]
WHERE  [cc].[Name] = 'DynamicSchema'
        AND [sa].[Name] = 'CoreCustomAttributes'
        AND [ca].[ContainerType] = 1
ORDER  BY [ss].Id DESC
```

**Review by khanaffan — CHANGES_REQUESTED** [review]

File: iModelCore/ECDb/ECDb/SchemaWriter.cpp [hunk 4]
```diff
@@ -134,6 +134,150 @@ DropSchemaResult SchemaWriter::DropSchema(Utf8StringCR schemaName, SchemaImportC
     }
     return DropSchemaResult(DropSchemaResult::Status::Success);
 }
+
+DropSchemaResult SchemaWriter::DropSchemas(T_Utf8StringVectorCR schemaNames, SchemaImportContext& ctx, bool logIssue)
+    {
```
Comment (ColinKerr) [inline iModelCore/ECDb/ECDb/SchemaWriter.cpp:139]: it concerns me that we have implementations for DropSchema and DropSchemas ... couldn't DropSchema be modified to call DropSchemas?

File: iModelCore/ECDb/ECDb/SchemaWriter.cpp (hunk 1 above)
Comment (RohitPtnkr1996) [reply iModelCore/ECDb/ECDb/SchemaWriter.cpp]: Done !

File: iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp (hunk 2 above)
Comment (RohitPtnkr1996) [reply iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp]: Done

File: iModelCore/ECDb/ECDb/SchemaWriter.cpp (hunk 4 above)
Comment (RohitPtnkr1996) [reply iModelCore/ECDb/ECDb/SchemaWriter.cpp:17]: Done

**Review by khanaffan — APPROVED** [review]

**Review by ColinKerr — APPROVED** [review]

File: iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp (hunk 3 above)
Comment (RohitPtnkr1996) [reply iModelCore/ECDb/ECDb/SchemaManagerDispatcher.cpp:1018]: Done. I've added code to sort the schemas so that the referenced schemas get dropped first.

**Review by nick4598 — APPROVED** [review]

**Pull request merged**